From aadf7f77b69794da9f082d28232faf88ffd53a3e Mon Sep 17 00:00:00 2001
From: Vadim Pasternak <vadimp@nvidia.com>
Date: Mon, 8 Nov 2021 09:07:52 +0000
Subject: [PATCH backport for v4.19 1/1] Squash to commit 16d59a4882b8 ("mlxsw:
 core: Add APIs for thermal sensor mapping")

Fix after PRM change of MTECR register bitmap layout.

Signed-off-by: Vadim Pasternak <vadimp@nvidia.com>
---
 drivers/net/ethernet/mellanox/mlxsw/core_env.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/mellanox/mlxsw/core_env.c b/drivers/net/ethernet/mellanox/mlxsw/core_env.c
index a73049dd1..b2ed87a78 100644
--- a/drivers/net/ethernet/mellanox/mlxsw/core_env.c
+++ b/drivers/net/ethernet/mellanox/mlxsw/core_env.c
@@ -325,7 +325,7 @@ int mlxsw_env_sensor_map_create(struct mlxsw_core *core,
 				struct mlxsw_env_gearbox_sensors_map *map)
 {
 	char mtecr_pl[MLXSW_REG_MTECR_LEN];
-	u16 last_sensor;
+	u16 last_sensor, offset;
 	int i, bit, err;
 
 	mlxsw_reg_mtecr_pack(mtecr_pl, slot_index);
@@ -344,8 +344,14 @@ int mlxsw_env_sensor_map_create(struct mlxsw_core *core,
 	if (!map->sensor_bit_map)
 		return -ENOMEM;
 
+	/* Sensors bitmap is size variable based on the last_sensor field and
+	 * in granularity of 32bits. Calculate an offset in payload buffer to
+	 * start from.
+	 */
+	offset = MLXSW_REG_MTECR_SENSOR_MAP_LEN * 8 - last_sensor - 1;
+	offset -= offset % 32;
 	for (bit = 0, i = 0; bit <= last_sensor && i < map->sensor_count; bit++) {
-		if (mlxsw_reg_mtecr_sensor_map_get(mtecr_pl, bit))
+		if (mlxsw_reg_mtecr_sensor_map_get(mtecr_pl, bit + offset))
 			map->sensor_bit_map[i++] = bit;
 	}
 
-- 
2.20.1

